File: sdk/typescript/src/namespaces/retention.ts
/**
 * Retention Namespace API
 *
 * Provides access to data retention policies and management.
 * Essential for compliance and data lifecycle management.
 */

/**
 * Retention policy definition
 */
export interface RetentionPolicy {
  id: string;
  name: string;
  description: string;
  retention_days: number;
  data_types: string[];
  action: 'archive' | 'delete' | 'anonymize';
  enabled: boolean;
  schedule?: string;
  last_executed?: string;
  next_execution?: string;
  created_at: string;
  updated_at: string;
}

/**
 * Expiring data item
 */
export interface ExpiringItem {
  id: string;
  type: string;
  expires_at: string;
  policy_id: string;
  metadata?: Record<string, unknown>;
}

/**
 * Policy execution result
 */
export interface ExecutionResult {
  policy_id: string;
  executed_at: string;
  items_processed: number;
  items_archived: number;
  items_deleted: number;
  items_anonymized: number;
  errors: string[];
}

/**
 * Internal client interface
 */
interface RetentionClientInterface {
  get<T>(path: string): Promise<T>;
  post<T>(path: string, body?: unknown): Promise<T>;
}

/**
 * Retention API namespace.
 *
 * Provides methods for retention policy management:
 * - List retention policies
 * - Execute policies manually
 * - View expiring data
 *
 * @example
 * ```typescript
 * const client = createClient({ baseUrl: 'https://api.aragora.ai', apiKey: 'your-key' });
 *
 * // List retention policies
 * const policies = await client.retention.listPolicies();
 *
 * // Execute a policy
 * const result = await client.retention.executePolicy('policy-123');
 *
 * // View expiring data
 * const expiring = await client.retention.getExpiring();
 * ```
 */
export class RetentionAPI {
  constructor(private client: RetentionClientInterface) {}

  /**
   * List retention policies.
   */
  async listPolicies(): Promise<{ policies: RetentionPolicy[] }> {
    return this.client.get('/api/v1/retention/policies');
  }

  /**
   * Execute a retention policy manually.
   */
  async executePolicy(policyId: string, options?: { dry_run?: boolean }): Promise<ExecutionResult> {
    return this.client.post(`/api/v1/retention/policies/${policyId}/execute`, options);
  }

  /**
   * Get data items that are expiring soon.
   */
  async getExpiring(options?: { days?: number; limit?: number }): Promise<{ items: ExpiringItem[]; total: number }> {
    const params = new URLSearchParams();
    if (options?.days) params.set('days', options.days.toString());
    if (options?.limit) params.set('limit', options.limit.toString());
    const query = params.toString() ? `?${params.toString()}` : '';
    return this.client.get(`/api/v1/retention/expiring${query}`);
  }
}

File: sdk/python/aragora/namespaces/retention.py
"""
Retention Namespace API

Provides access to data retention policies and management.
Essential for compliance and data lifecycle management.

Features:
- List retention policies
- Execute policies manually
- View expiring data
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any, Literal

if TYPE_CHECKING:
    from ..client import AragoraAsyncClient, AragoraClient


RetentionAction = Literal["archive", "delete", "anonymize"]


class RetentionAPI:
    """
    Synchronous Retention API.

    Provides methods for retention policy management:
    - List retention policies
    - Execute policies manually
    - View expiring data

    Example:
        >>> client = AragoraClient(base_url="https://api.aragora.ai", api_key="...")
        >>> policies = client.retention.list_policies()
        >>> expiring = client.retention.get_expiring(days=7)
    """

    def __init__(self, client: AragoraClient):
        self._client = client

    def list_policies(self) -> dict[str, Any]:
        """
        List retention policies.

        Returns:
            Dict with policies list containing:
            - id: Policy ID
            - name: Policy name
            - description: Policy description
            - retention_days: Days to retain
            - data_types: Affected data types
            - action: Action (archive/delete/anonymize)
            - enabled: Whether enabled
            - schedule: Execution schedule
        """
        return self._client.request("GET", "/api/v1/retention/policies")

    def execute_policy(
        self,
        policy_id: str,
        dry_run: bool = False,
    ) -> dict[str, Any]:
        """
        Execute a retention policy manually.

        Args:
            policy_id: The policy ID
            dry_run: If True, simulate without making changes

        Returns:
            Dict with execution results:
            - policy_id: Executed policy
            - executed_at: Execution timestamp
            - items_processed: Total items
            - items_archived: Archived count
            - items_deleted: Deleted count
            - items_anonymized: Anonymized count
            - errors: Any errors
        """
        return self._client.request(
            "POST",
            f"/api/v1/retention/policies/{policy_id}/execute",
            json={"dry_run": dry_run} if dry_run else None,
        )

    def get_expiring(
        self,
        days: int | None = None,
        limit: int | None = None,
    ) -> dict[str, Any]:
        """
        Get data items that are expiring soon.

        Args:
            days: Number of days to look ahead
            limit: Maximum items to return

        Returns:
            Dict with:
            - items: List of expiring items
            - total: Total count
        """
        params: dict[str, Any] = {}
        if days:
            params["days"] = days
        if limit:
            params["limit"] = limit
        return self._client.request(
            "GET", "/api/v1/retention/expiring", params=params if params else None
        )


class AsyncRetentionAPI:
    """
    Asynchronous Retention API.

    Example:
        >>> async with AragoraAsyncClient(base_url="https://api.aragora.ai") as client:
        ...     policies = await client.retention.list_policies()
        ...     result = await client.retention.execute_policy("policy-123", dry_run=True)
    """

    def __init__(self, client: AragoraAsyncClient):
        self._client = client

    async def list_policies(self) -> dict[str, Any]:
        """List retention policies."""
        return await self._client.request("GET", "/api/v1/retention/policies")

    async def execute_policy(
        self,
        policy_id: str,
        dry_run: bool = False,
    ) -> dict[str, Any]:
        """Execute a retention policy manually."""
        return await self._client.request(
            "POST",
            f"/api/v1/retention/policies/{policy_id}/execute",
            json={"dry_run": dry_run} if dry_run else None,
        )

    async def get_expiring(
        self,
        days: int | None = None,
        limit: int | None = None,
    ) -> dict[str, Any]:
        """Get data items that are expiring soon."""
        params: dict[str, Any] = {}
        if days:
            params["days"] = days
        if limit:
            params["limit"] = limit
        return await self._client.request(
            "GET", "/api/v1/retention/expiring", params=params if params else None
        )
